### ISSUE TITLE
RTX 3060 12GB Benchmarking

### ISSUE THREAD

**Issue by 1aienthusiast** [turboderp/exllama#6]
model: llama-13B-4bit-128g

exllama:
```
(exllama) user@debian:~/AI/exllama$ python test_benchmark_inference.py -d ~/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/ -p
 -- Loading model
 -- Tokenizer: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/tokenizer.model
 -- Model config: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/config.json
 -- Model: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/llama-13b-4bit-128g.safetensors
 -- Sequence length: 2048
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched', 'perf']
 ** Time, Load model: 1.57 seconds
 -- Groupsize (inferred): 128
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 6,683.17 MB
 -- Inference, first pass.
 ** Time, Inference: 2.08 seconds
 ** Speed: 923.57 tokens/second
 -- Generating 128 tokens...
 ** Speed: 22.04 tokens/second
 ** VRAM, Inference: [cuda:0] 2,291.67 MB
 ** VRAM, Total: [cuda:0] 8,974.84 MB
```

ooba's webui:
streaming on:
```
(textgen) user@debian:~/AI/2oobabooga/text-generation-webui$ python3.10 server.py --wbits 4 --model llama-13b-4bit-128g --groups 128 --model_type llama
INFO:Gradio HTTP request redirected to localhost :)
bin /home/user/miniconda3/envs/textgen/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cuda117.so
INFO:Loading llama-13b-4bit-128g...
INFO:Found the following quantized model: models/llama-13b-4bit-128g/llama-13b-4bit-128g.safetensors
INFO:Loaded the model in 2.65 seconds.
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Output generated in 9.12 seconds (21.81 tokens/s, 199 tokens, context 4, seed 989197438)
Output generated in 8.57 seconds (23.22 tokens/s, 199 tokens, context 4, seed 26472177)

```

no stream:
```
(textgen) user@debian:~/AI/2oobabooga/text-generation-webui$ python3.10 server.py --wbits 4 --model llama-13b-4bit-128g --groups 128 --model_type llama --no-stream
INFO:Gradio HTTP request redirected to localhost :)
bin /home/user/miniconda3/envs/textgen/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cuda117.so
INFO:Loading llama-13b-4bit-128g...
INFO:Found the following quantized model: models/llama-13b-4bit-128g/llama-13b-4bit-128g.safetensors
INFO:Loaded the model in 2.48 seconds.
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Output generated in 5.17 seconds (24.74 tokens/s, 128 tokens, context 4, seed 250438644)
Output generated in 4.57 seconds (28.02 tokens/s, 128 tokens, context 4, seed 1203371762)
Output generated in 4.80 seconds (26.65 tokens/s, 128 tokens, context 4, seed 484445001)
```

**Comment by turboderp** [comment]
Thank you.

It's not really the way to compare, though, since ExLlama starts with a context of 1920 tokens (that's the 923.57 t/s in your output) and then generates 128 individual tokens from there (i.e. it tests worst-case performance.)

The way the benchmark script works, you'd want to call it with `-l 132` for an equivalent test, or pass Oogabooga a 1920-token context.

I do plan on doing some more tuning for 30-series cards also at some point.

**Comment by 1aienthusiast** [comment]
did the testing again with `-l 132`:

```
(exllama) user@debian:~/AI/exllama$ python test_benchmark_inference.py -l 132 -d ~/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/ -p
 -- Loading model
 -- Tokenizer: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/tokenizer.model
 -- Model config: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/config.json
 -- Model: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/llama-13b-4bit-128g.safetensors
 -- Sequence length: 132
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched', 'perf']
 ** Time, Load model: 1.55 seconds
 -- Groupsize (inferred): 128
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 6,682.23 MB
 -- Inference, first pass.
 ** Time, Inference: 0.50 seconds
 ** Speed: 8.06 tokens/second
 -- Generating 128 tokens...
 ** Speed: 28.56 tokens/second
 ** VRAM, Inference: [cuda:0] 123.03 MB
 ** VRAM, Total: [cuda:0] 6,805.26 MB
(exllama) user@debian:~/AI/exllama$ python test_benchmark_inference.py -l 132 -d ~/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/ -p
 -- Loading model
 -- Tokenizer: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/tokenizer.model
 -- Model config: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/config.json
 -- Model: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/llama-13b-4bit-128g.safetensors
 -- Sequence length: 132
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched', 'perf']
 ** Time, Load model: 1.56 seconds
 -- Groupsize (inferred): 128
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 6,682.23 MB
 -- Inference, first pass.
 ** Time, Inference: 0.50 seconds
 ** Speed: 7.95 tokens/second
 -- Generating 128 tokens...
 ** Speed: 28.56 tokens/second
 ** VRAM, Inference: [cuda:0] 123.03 MB
 ** VRAM, Total: [cuda:0] 6,805.26 MB
(exllama) user@debian:~/AI/exllama$ python test_benchmark_inference.py -l 132 -d ~/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/ -p
 -- Loading model
 -- Tokenizer: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/tokenizer.model
 -- Model config: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/config.json
 -- Model: /home/user/AI/2oobabooga/text-generation-webui/models/llama-13b-4bit-128g/llama-13b-4bit-128g.safetensors
 -- Sequence length: 132
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched', 'perf']
 ** Time, Load model: 1.59 seconds
 -- Groupsize (inferred): 128
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 6,682.23 MB
 -- Inference, first pass.
 ** Time, Inference: 0.50 seconds
 ** Speed: 7.99 tokens/second
 -- Generating 128 tokens...
 ** Speed: 28.61 tokens/second
 ** VRAM, Inference: [cuda:0] 123.03 MB
 ** VRAM, Total: [cuda:0] 6,805.26 MB
```

i also noticed that the parameters preset has effects on performance in ooba's webui, so im assuming that parameters could have effect on your benchmark aswell

```
(textgen) user@debian:~/AI/2oobabooga/text-generation-webui$ python3.10 server.py --wbits 4 --model llama-13b-4bit-128g --groups 128 --model_type llama --no-stream
INFO:Gradio HTTP request redirected to localhost :)
bin /home/user/miniconda3/envs/textgen/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cuda117.so
INFO:Loading llama-13b-4bit-128g...
INFO:Found the following quantized model: models/llama-13b-4bit-128g/llama-13b-4bit-128g.safetensors
INFO:Loaded the model in 2.55 seconds.
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Output generated in 5.22 seconds (24.51 tokens/s, 128 tokens, context 4, seed 551704178)
Output generated in 4.61 seconds (27.74 tokens/s, 128 tokens, context 4, seed 1291148787)
Output generated in 4.53 seconds (28.24 tokens/s, 128 tokens, context 4, seed 1747752802)
Output generated in 4.57 seconds (28.03 tokens/s, 128 tokens, context 4, seed 1478802079)
Output generated in 4.58 seconds (27.96 tokens/s, 128 tokens, context 4, seed 466913621)
##########CHANGED PRESET FROM LLAMA-PRECISE TO NOVELAI-STORYWRITER
Output generated in 4.44 seconds (28.83 tokens/s, 128 tokens, context 4, seed 235330997) 
Output generated in 4.46 seconds (28.70 tokens/s, 128 tokens, context 4, seed 1171444620)
Output generated in 4.57 seconds (28.02 tokens/s, 128 tokens, context 4, seed 145639840)
Output generated in 1.29 seconds (27.11 tokens/s, 35 tokens, context 4, seed 1551268189)
Output generated in 4.45 seconds (28.77 tokens/s, 128 tokens, context 4, seed 1532859112)
Output generated in 4.57 seconds (28.00 tokens/s, 128 tokens, context 4, seed 1669820113)


```

**Comment by 1aienthusiast** [comment]
llama-7b-4bit (13B OOM's on ooba with 1800+ tokens)

exllama without -l:

```
(exllama) user@debian:~/AI/exllama$ python test_benchmark_inference.py -d ~/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/ -p
 -- Loading model
 -- Tokenizer: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/tokenizer.model
 -- Model config: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/config.json
 -- Model: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/llama-7b-4bit.safetensors
 -- Sequence length: 2048
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched', 'perf']
 ** Time, Load model: 1.20 seconds
 -- Groupsize (inferred): None
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 3,390.76 MB
 -- Inference, first pass.
 ** Time, Inference: 1.35 seconds
 ** Speed: 1423.20 tokens/second
 -- Generating 128 tokens...
 ** Speed: 36.62 tokens/second
 ** VRAM, Inference: [cuda:0] 1,585.17 MB
 ** VRAM, Total: [cuda:0] 4,975.93 MB
(exllama) user@debian:~/AI/exllama$ python test_benchmark_inference.py -d ~/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/ -p
 -- Loading model
 -- Tokenizer: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/tokenizer.model
 -- Model config: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/config.json
 -- Model: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/llama-7b-4bit.safetensors
 -- Sequence length: 2048
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched', 'perf']
 ** Time, Load model: 1.17 seconds
 -- Groupsize (inferred): None
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 3,390.76 MB
 -- Inference, first pass.
 ** Time, Inference: 1.31 seconds
 ** Speed: 1466.86 tokens/second
 -- Generating 128 tokens...
 ** Speed: 37.41 tokens/second
 ** VRAM, Inference: [cuda:0] 1,585.17 MB
 ** VRAM, Total: [cuda:0] 4,975.93 MB
(exllama) user@debian:~/AI/exllama$ python test_benchmark_inference.py -d ~/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/ -p
 -- Loading model
 -- Tokenizer: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/tokenizer.model
 -- Model config: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/config.json
 -- Model: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/llama-7b-4bit.safetensors
 -- Sequence length: 2048
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched', 'perf']
 ** Time, Load model: 1.16 seconds
 -- Groupsize (inferred): None
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 3,390.76 MB
 -- Inference, first pass.
 ** Time, Inference: 1.31 seconds
 ** Speed: 1467.22 tokens/second
 -- Generating 128 tokens...
 ** Speed: 37.32 tokens/second
 ** VRAM, Inference: [cuda:0] 1,585.17 MB
 ** VRAM, Total: [cuda:0] 4,975.93 MB
(exllama) user@debian:~/AI/exllama$ python test_benchmark_inference.py -d ~/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/ -p
 -- Loading model
 -- Tokenizer: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/tokenizer.model
 -- Model config: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/config.json
 -- Model: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/llama-7b-4bit.safetensors
 -- Sequence length: 2048
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched', 'perf']
 ** Time, Load model: 1.15 seconds
 -- Groupsize (inferred): None
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 3,390.76 MB
 -- Inference, first pass.
 ** Time, Inference: 1.31 seconds
 ** Speed: 1471.11 tokens/second
 -- Generating 128 tokens...
 ** Speed: 37.38 tokens/second
 ** VRAM, Inference: [cuda:0] 1,585.17 MB
 ** VRAM, Total: [cuda:0] 4,975.93 MB
(exllama) user@debian:~/AI/exllama$ python test_benchmark_inference.py -d ~/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/ -p
 -- Loading model
 -- Tokenizer: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/tokenizer.model
 -- Model config: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/config.json
 -- Model: /home/user/AI/2oobabooga/text-generation-webui/models/llama-7b-4bit/llama-7b-4bit.safetensors
 -- Sequence length: 2048
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched', 'perf']
 ** Time, Load model: 1.16 seconds
 -- Groupsize (inferred): None
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 3,390.76 MB
 -- Inference, first pass.
 ** Time, Inference: 1.31 seconds
 ** Speed: 1467.59 tokens/second
 -- Generating 128 tokens...
 ** Speed: 37.38 tokens/second
 ** VRAM, Inference: [cuda:0] 1,585.17 MB
 ** VRAM, Total: [cuda:0] 4,975.93 MB
(exllama) user@debian:~/AI/exllama$
```

ooba's webui:

```
(textgen) user@debian:~/AI/2oobabooga/text-generation-webui$ python3.10 server.py --wbits 4 --model llama-7b-4bit  --model_type llama --no-stream
INFO:Gradio HTTP request redirected to localhost :)
bin /home/user/miniconda3/envs/textgen/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cuda117.so
INFO:Loading llama-7b-4bit...
INFO:Found the following quantized model: models/llama-7b-4bit/llama-7b-4bit.safetensors
INFO:Loaded the model in 4.42 seconds.
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Output generated in 55.88 seconds (34.36 tokens/s, 1920 tokens, context 4, seed 1806647125)
Output generated in 7.07 seconds (18.11 tokens/s, 128 tokens, context 1920, seed 1706572121)
Output generated in 7.08 seconds (18.07 tokens/s, 128 tokens, context 1920, seed 1478787660)

```

indeed, seems to be 2x faster with high context, very impressive!

great job!

**Comment by turboderp** [comment]
It could get faster still. It's early days. :)

I'm not sure what the "parameters preset" is in text-generation-webui, but I doubt it would have an impact for ExLlama. If it's just sampling/generation parameters, then the difference would be microscopic. In fact the benchmark doesn't even do any actual sampling, it just produces logits.

I guess it also doesn't stream to a UI and that does seem to be an issue with text-generation-webui. Is it normal that streaming causes it to slow down?

**Comment by Fairfax-Mooresby** [comment]
Given the 5t/s difference between my 3060 (~15t/s) and 1aienthusiast's(20t/s) I suspect CPU or other bottleneck. I'm on an ancient Xeon e5 V2 , which has no AVX2 and tops out at around 3.5Ghz, and the board is PCIe v3 (barely, it's 10 years old)

exllama is significantly faster for me than Ooba with multi-gpu layering on 33b, testing a 'chat' and allowing some context to build up exllama is about twice as fast.

**Comment by 1aienthusiast** [comment]
> It could get faster still. It's early days. :)
> 
> I'm not sure what the "parameters preset" is in text-generation-webui, but I doubt it would have an impact for ExLlama. If it's just sampling/generation parameters, then the difference would be microscopic. In fact the benchmark doesn't even do any actual sampling, it just produces logits.
> 
> I guess it also doesn't stream to a UI and that does seem to be an issue with text-generation-webui. Is it normal that streaming causes it to slow down?

yeah its normal for streaming to cause slowdowns, atleast in ooba's webui



> Given the 5t/s difference between my 3060 (~15t/s) and 1aienthusiast's(20t/s) I suspect CPU or other bottleneck. I'm on an ancient Xeon e5 V2 , which has no AVX2 and tops out at around 3.5Ghz, and the board is PCIe v3 (barely, it's 10 years old)
> 
> exllama is significantly faster for me than Ooba with multi-gpu layering on 33b, testing a 'chat' and allowing some context to build up exllama is about twice as fast.

im using an i5 12400f
